allow --experimental_remote_cache_ttl=server to request remote cache entries that live as long as the Bazel server

dws opened the pull request:
Under this flag, Bazel will retain remote execution records in its
persistent action cache for as long as the server runs.  To this end, we
introduce a new FileArtifactValue expirationTime sentinel value of -2.

Note that we bump the VERSION of CompactPersistentActionCache.  This
way, the right thing will happen if an earlier version of Bazel that
does not have the logic to selectively clear the expirationTime ==
-2 entries inherits a persistent action cache from a later version of
Bazel.

In order to see if your CompactPersistentActionCache has server-duration
entries, you can see if:

    bazel dump --action_cache | grep expirationTime=-2

finds anything.

fmeum commented on src/main/java/com/google/devtools/build/lib/remote/options/CommonRemoteOptions.java:
```diff
@@ -64,6 +68,10 @@ public static class RemoteDurationConverter extends Converter.Contextless<Durati
     @Override
     public Duration convert(String input) throws OptionsParsingException {
+      /* We recognize the magic value SERVER_KEYWORD as a way to specify server lifetime ttl. */
+      if (input.equals(SERVER_KEYWORD)) {
```
Please move this to a dedicated new converter class, possibly by subclassing this one. The keyword just doesn't make sense for other duration flags such as `--remote_retry_max_delay`.

dws replied:
Agreed -- done.

fmeum commented on src/main/java/com/google/devtools/build/lib/runtime/BlazeWorkspace.java:
```diff
@@ -289,6 +298,22 @@ public ActionCache getOrLoadPersistentActionCache(Reporter reporter) throws IOEx
+        try (AutoProfiler p = profiledAndLogged("pruning server-lifetime entries from action cache", ProfilerTask.INFO)) {
+          actionCache.removeIf(
+            entry ->
+            entry.getOutputFiles().values().stream().filter(e -> e.getExpireAtEpochMilli() == RemoteFileArtifactValue.SERVER_EXPIRATION_SENTINEL).count() > 0 ||
+            entry.getOutputTrees().values().stream().filter(
+              tv -> tv.childValues().values().stream().filter(e -> e.getExpireAtEpochMilli() == RemoteFileArtifactValue.SERVER_EXPIRATION_SENTINEL).count() > 0
+            ).count() > 0
```
```suggestion
                  Stream.concat(
                          entry.getOutputFiles().values().stream(),
                          entry.getOutputTrees().values().stream()
                              .flatMap(tv -> tv.childValues().values().stream()))
                      .anyMatch(e -> e.getExpirationTime() == RemoteFileArtifactValue.SERVER_EXPIRATION_SENTINEL));
```

dws replied:
Great suggestion, thanks!

fmeum commented on src/main/java/com/google/devtools/build/lib/actions/FileArtifactValue.java:
```diff
@@ -558,6 +561,9 @@ public static RemoteFileArtifactValue create(
         long expireAtEpochMilli,
         @Nullable PathFragment materializationExecPath) {
+      if (expireAtEpochMilli == SERVER_EXPIRATION_SENTINEL) {
+        return new RemoteFileArtifactValueWithServerExpiration(digest, size, locationIndex, materializationExecPath);
```
This will require some rebasing.

dws replied:
Indeed it did.

dws committed:
allow --experimental_remote_cache_ttl=server to request remote cache entries that live as long as the Bazel server

Under this flag, Bazel will retain remote execution records in its
persistent action cache for as long as the server runs.  To this end, we
introduce a new FileArtifactValue expirationTime sentinel value of -2.

Note that we bump the VERSION of CompactPersistentActionCache.  This
way, the right thing will happen if an earlier version of Bazel that
does not have the logic to selectively clear the expirationTime ==
-2 entries inherits a persistent action cache from a later version of
Bazel.

In order to see if your CompactPersistentActionCache has server-duration
entries, you can see if:

    bazel dump --action_cache | grep expirationTime=-2

finds anything.

fmeum commented on src/main/java/com/google/devtools/build/lib/runtime/BlazeWorkspace.java:
```diff
@@ -395,6 +406,23 @@ public ActionCache getOrLoadPersistentActionCache(Reporter reporter) throws IOEx
+      if (clearActionCacheServerLifetimeEntries) {
+        // We only exercise this code at most once during the lifetime of the server.
+        clearActionCacheServerLifetimeEntries = false;
+        try (AutoProfiler p = profiledAndLogged("pruning server-lifetime entries from action cache", ProfilerTask.INFO)
```
How long does this take on a larger repo? It's a cost paid even by someone who doesn't use the new feature, so we need to be careful.

dws committed:
add //src/test/java/com/google/devtools/build/lib/remote:ServerLifetimeIntegrationTest

Per reviewer feedback, we here add an integration test to
exercise --experimental_remote_cache_ttl=server.

tjgq commented:
> @dws Taking a step back, what do you think of the approach described in [#26140 (comment)](https://github.com/bazelbuild/bazel/issues/26140#issuecomment-2909518732)? I guess that the main motivation behind `--experimental_remote_cache_ttl=server` is to avoid the eager invalidation performed by Bazel, which that change would avoid in a different way?

+1. I'd prefer to avoid a full scan and instead defer invalidation to the time at which a particular entry is accessed.

Even if the idea described in https://github.com/bazelbuild/bazel/issues/26140#issuecomment-2909518732 is insufficient to achieve the desired outcome, we could still tie individual action cache entries to the lifetime of a particular server, by generating a random or auto-incremented value every time the server is started, and storing it instead of an expiration time in individual entries. But let's try simpler solutions first, please.

dws commented:
@bazelbuild/triage

I would like to resume work on this.

fmeum commented:
@dws I'm working on removing the lease service in favor of action rewinding, with which Bazel can trust all remote data and recover if it turns out to have been evicted. I think that this would address your use case by default. Bazel 9.3.0 should already have this behavior.

dws commented:
@fmeum Thanks for letting me know -- I'm glad to hear it!

One case where we'd like the right thing to happen automatically is when switching between different `--remote_executor` settings, or having a given `--remote_executor` setting change its meaning due to a DNS update.

fmeum commented:
Could you test this case with the latest commit on the release-9.3.0 branch and/or file an issue? I haven't considered this before, but would expect it to work with a small performance penalty on the first build after the switch.
